src/puppetlabs/puppetdb/mq_listener.clj

```diff
@@ -133,174 +156,99 @@
+(defn mark-both-metrics!
+  "Calls `mark!` on the global and command specific metric for `k`"
+  [command version k]
+  (mark! (global-metric k))
+  (mark! (cmd-metric command version k)))
+
+(defn update-both-metrics!
+  "Calls `update!` on the global and command specific metric for `k`"
+  [command version k v]
+  (update! (global-metric k) v)
+  (update! (cmd-metric command version k) v))
+
+(defn call-with-command-metrics
+  "Invokes `f` including the related metrics updates"
+  [command version retries f]
+  (create-metrics-for-command! command version)
+
+  (mark-both-metrics! command version :seen)
+  (update-both-metrics! command version :retry-counts retries)
+
+  (mutils/multitime!
+   [(global-metric :processing-time)
+    (cmd-metric command version :processing-time)]
+
+   (let [command-result (f)]
+     (mark-both-metrics! command version :processed)
+     command-result)))
+
+;; Primarily a test hook
+(defn discard-message [message q dlo]
+  (dlo/discard-cmdref message q dlo))
+
+(defn message-handler
+  "Processes the message via (process-message msg), retrying messages
+  that fail via (delay-message msg), and discarding messages that have
+  fatal errors or have exceeded their maximum allowed attempts."
+  [q dlo delay-message process-message]
+  (fn [{:keys [certname command version delete? id] :as cmdref}]
+    (try+
+     ;; If the message is a delete?, there's no need to parse it
+     ;; below, it's only going to be removed
+     (if delete?
+       (do
+         (process-message cmdref)
+         (queue/ack-command q {:entry (queue/cmdref->entry cmdref)})
+         (update-counter! :depth command version dec!)
+         (update-counter! :invalidated command version dec!))
+       (let [cmd (queue/cmdref->cmd q cmdref)
+             retries (count (:attempts cmdref))]
+         (try+
+           (call-with-command-metrics command version retries
+                                      #(process-message cmd))
+          (queue/ack-command q cmd)
+          (update-counter! :depth command version dec!)
+
+          (catch fatal? obj
+            (mark! (global-metric :fatal))
+            (let [ex (:cause obj)]
+              (log/error (:wrapper &throw-context) (i18n/trs "[{0}] [{1}] Fatal error on attempt {2} for {3}" id command retries certname))
+              (-> cmd
+                  (queue/cons-attempt ex)
+                  (discard-message q dlo))))
+          (catch Exception _
+            (let [ex (:throwable &throw-context)
+                  log-str (i18n/trs "[{0}] [{1}] Retrying after attempt {2} for {3}, due to: {4}"
+                                    id command retries certname ex)]
+              (mark-both-metrics! command version :retried)
+              (cond
+                (< retries 4)
+                (do
+                  (log/debug ex log-str)
+                  (-> cmd (queue/cons-attempt ex) delay-message))
+
+                (< retries maximum-allowable-retries)
```

Comment:
This fixes a bug currently in master. In master, the last attempt is skipped. This patch will try up to `maximum-allowable-retries` then fail (else clause below)